### PR TITLE
pip: Refactor how sources are generated 

### PR DESCRIPTION
**Pull request by A6GibKm**
DONE: ~~Find a way to list the sources files only once.~~
 with this PR the ALL sources will be listed once for each package. One possibility is to, instead of doing one module per package, do a single module with all packages. This would drastically decrease the length of the output file, even compared with `flatpak-pip-generator` before this PR, but we lose the ability to take off from the last built module in `flatpak-builder`.

### CODE REVIEW

**Comment by A6GibKm** [comment]
Since `file` sources are downloaded only once when running `flatpak-builder`, the only drawback to this are:
* vcs modules will download still download more than once
* The output file will be harder to read due to the sources list being, probably gigantic.

**Comment by A6GibKm** [comment]
This commit does `pip download package` on all packages (yes, packages can be downloaded up to 3 times, but most of the time they are on cache) to generate a dependency list and only list sources according to it. A small benchmark on 
```
pyaes>=0.1a1
ecdsa>=0.14
qrcode
protobuf>=3.12
dnspython<2.0
qdarkstyle<2.9
aiorpcx>=0.18,<0.19
aiohttp>=3.3.0,<4.0.0
aiohttp_socks>=0.3
certifi
bitstring
attrs>=19.2.0
```
Shows that even if `flatpak-pip-generator` does more calls of `pip download` the runtime %50 of what it was, and the output file is ~14% shorter, and only installs the correct versions.

File: pip/flatpak-pip-generator [hunk 1]
```diff
@@ -74,12 +74,45 @@ def get_file_hash(filename: str) -> str:
...
+            if '--hash' in line:
+                use_hash = '--require-hashes'
+                package_name, hash_requirements = parse_hashes(line)
+                reqs[i] = package_name
+        reqs_as_str = "\n".join(reqs)
+        packages = list(requirements.parse(reqs_as_str))
+        requirements_file = opts.requirements_file
 elif opts.packages:
     packages = list(requirements.parse('\n'.join(opts.packages)))
+    requirements_file = '/tmp/flatpak-pip-generator-requirements.txt'
```
Comment (TingPing) [inline pip/flatpak-pip-generator:116]: Use the `tempfile` module.

File: pip/flatpak-pip-generator [hunk 2]
```diff
@@ -74,12 +74,45 @@ def get_file_hash(filename: str) -> str:
...
+        hash_requirements.append([hash_algorithm.lower(), hash_value])
+    return name, hash_requirements
+
+
 packages = []
+use_hash = None
 if opts.requirements_file and os.path.exists(opts.requirements_file):
     with open(opts.requirements_file, 'r') as req_file:
-        packages = list(requirements.parse(req_file))
+        reqs = list(parse_continuation_lines(req_file))
```
Comment (TingPing) [inline pip/flatpak-pip-generator:105]: Why `list()`

File: pip/flatpak-pip-generator [hunk 3]
```diff
@@ -74,12 +74,45 @@ def get_file_hash(filename: str) -> str:
...
+        for i, line in enumerate(reqs):
+            if '--hash' in line:
+                use_hash = '--require-hashes'
```
Comment (TingPing) [inline pip/flatpak-pip-generator:108]: Maybe turn this into a `bool` and then just insert the string below.

File: pip/flatpak-pip-generator [hunk 4]
```diff
@@ -74,12 +74,45 @@ def get_file_hash(filename: str) -> str:
...
+                package_name, hash_requirements = parse_hashes(line)
+                reqs[i] = package_name
+        reqs_as_str = "\n".join(reqs)
+        packages = list(requirements.parse(reqs_as_str))
+        requirements_file = opts.requirements_file
 elif opts.packages:
     packages = list(requirements.parse('\n'.join(opts.packages)))
+    requirements_file = '/tmp/flatpak-pip-generator-requirements.txt'
+    with open(requirements_file, 'w') as req_file:
+        req_file.writelines('\n'.join(opts.packages))
```
Comment (TingPing) [inline pip/flatpak-pip-generator:118]: The point of `writelines` is it takes an iterable and you don't need to concat yourself?

File: pip/flatpak-pip-generator [hunk 5]
```diff
@@ -74,12 +74,45 @@ def get_file_hash(filename: str) -> str:
...
+
+
+def parse_hashes(line):
+    name = line.split(" ")[0]
+    hash_requirements = []
+    while "--hash=" in line:
+        split = line.split("--hash=", 1)
+        hash_algorithm = split[1].split(":", 1)[0]
+        hash_value = split[1].split(":", 1)[1].split(" ", 1)[0]
+        line = line.replace('--hash=' + hash_algorithm + ":" + hash_value, '')
```
Comment (TingPing) [inline pip/flatpak-pip-generator:96]: Double quotes are used

File: pip/flatpak-pip-generator (hunk 1 above)
Comment (A6GibKm) [reply pip/flatpak-pip-generator:116]: The temp file does not live enough, I will try it again tho.

Comment (A6GibKm) [reply pip/flatpak-pip-generator]: Fixed, there is a `delete=False` parameter.

File: pip/flatpak-pip-generator (hunk 3 above)
Comment (A6GibKm) [reply pip/flatpak-pip-generator]: Done.

File: pip/flatpak-pip-generator (hunk 4 above)
Comment (A6GibKm) [reply pip/flatpak-pip-generator]: Yes, but it does not add the newlines. But as you point, `write` is more apt.

File: pip/flatpak-pip-generator (hunk 2 above)
Comment (A6GibKm) [reply pip/flatpak-pip-generator:105]: I am not happy with it either. What I do is to iter over `reqs` and change its contents (via `reqs[i]=foo`). Later `list()` is used again for the `packages` iterable so `len(packages) == 1` can be called without consuming the iterable, I haven't tough of a more graceful way of handling it.

File: pip/flatpak-pip-generator [hunk 6]
```diff
@@ -74,15 +74,36 @@ def get_file_hash(filename: str) -> str:
...
 if opts.requirements_file and os.path.exists(opts.requirements_file):
     with open(opts.requirements_file, 'r') as req_file:
-        packages = list(requirements.parse(req_file))
+        reqs = parse_continuation_lines(req_file)
+        reqs_as_str = '\n'.join([r.split('--hash')[0] for r in reqs])
+        packages = list(requirements.parse(reqs_as_str))
+        requirements_file = opts.requirements_file
 elif opts.packages:
     packages = list(requirements.parse('\n'.join(opts.packages)))
+    with tempfile.NamedTemporaryFile('w', delete=False) as req_file:
```
Comment (TingPing) [inline pip/flatpak-pip-generator:98]: This does mean you leave temp files around forever which isn't ideal.

Comment (TingPing) [reply pip/flatpak-pip-generator:98]: Oh at the end you had a deletion check. You just need to store this object around and let it get deleted.

Comment (A6GibKm) [reply pip/flatpak-pip-generator:98]: Does this commit resolve the issue?

File: pip/flatpak-pip-generator [hunk 7]
```diff
@@ -311,9 +311,8 @@ for package in packages:
     else:
         modules.append(module)
 
-if requirements_file == '/tmp/flatpak-pip-generator-requirements.txt':
-    if os.path.exists(requirements_file):
-        os.remove(requirements_file)
+if not opts.requirements_file and os.path.exists(requirements_file):
+    os.remove(requirements_file)
```
Comment (TingPing) [inline pip/flatpak-pip-generator:315]: This is a race condition. Just always call `os.remove()` and handle it failing as normal.

**Comment by TingPing** [comment]
Thanks!

**Pull request merged**